product_kits_inventory/product.py

```diff
@@ -0,0 +1,313 @@
+                          'what': ('in',)})
+            if f == 'outgoing_qty':
+                c.update({'states': ('confirmed', 'waiting', 'assigned'),
+                          'what': ('out',)})
+            stock = self.get_product_available(cr, uid, ids, context=c)
+            for id in ids:
+                res[id][f] = stock.get(id, 0.0)
+                product_obj = self.pool['product.product'].browse(cr, uid, id)
+                res[id]['product_uom'] = product_obj.uom_id
+        return res
+
+    # get sale quotation of this product.
+    def _get_sale_quotation_qty(self, cr, uid, product_id, context=None):
+        '''get all qty of the product in all sale quotations (draft, sent)'''
+        total = 0.0
+        sol_obj = self.pool['sale.order.line']
+        product_uom_obj = self.pool['product.uom']
+        domain = [('state', 'in', ('draft', False, None)),
+                  ('product_id', '=', product_id)]
+        sol_ids = sol_obj.search(cr, uid, domain)
+        for sol_id in sol_ids:
+            solobj = sol_obj.browse(cr, uid, sol_id)
+            if not solobj.product_id:
+                continue
+            total += product_uom_obj._compute_qty(cr, uid,
+                                                  solobj.product_uom.id,
+                                                  solobj.product_uom_qty,
+                                                  solobj.product_id.uom_id.id)
+        return total
+
+    def _get_qty_from_children(
+            self, cr, uid, child_product_qtys, child_product_res, field_name):
+        def qty_div(product_total_qty, component_qty):
+            product_uom_obj = self.pool['product.uom']
+            kit_number = product_uom_obj._compute_qty(
+                cr, uid, component_qty[1]['mrp_product_uom'].id,
+                component_qty[1]['mrp_product_qty'],
+                component_qty[1]['uom'].id)
+            amout = product_uom_obj._compute_qty(
+                cr, uid, component_qty[1]['product_uom'].id,
+                component_qty[1]['product_qty'],
+                component_qty[1]['uom'].id)
+            com_number = amout / kit_number
+            com_total = product_uom_obj._compute_qty(
+                cr, uid, product_total_qty[1]['product_uom'].id,
+                product_total_qty[1].get(field_name),
+                component_qty[1]['uom'].id)
+            number = com_total / com_number
+            return int(number)
+
+        # when the bom has no components
+        if not child_product_res:
+            return 0
+        return min(map(qty_div, child_product_qtys.iteritems(),
+                   child_product_res.iteritems()))
+
+    def _is_kit(self, cr, uid, ids, fields=None, args=False, context=None):
+        '''see if this product is Kit or not'''
+        res = {}
+        for product_record in self.browse(cr, uid, ids, context=context):
+            res[product_record.id] = False
```

Comment:
You could just use a search with 'count=True' statement instead of browsing products, then BoM.

e.g. :

    res[product_record.id] = self.pool.get('mrp.bom').search(cr, uid, [('product id', '=', product_record.id), ('type', '=', 'phantom')], count=True, context=context) and True or False